src/redisearch_rs/result_processor/src/ffi.rs

```diff
@@ -0,0 +1,236 @@
+    pub const RS_RESULT_OK: Self = Self(0);
+    /// Result is empty, and the last result has already been returned.
+    pub const RS_RESULT_EOF: Self = Self(1);
+    /// Execution paused due to rate limiting (or manual pause from ext. thread??)
+    pub const RS_RESULT_PAUSED: Self = Self(2);
+    /// Execution halted because of timeout
+    pub const RS_RESULT_TIMEDOUT: Self = Self(3);
+    /// Aborted because of error. The QueryState (parent->status) should have
+    /// more information.
+    pub const RS_RESULT_ERROR: Self = Self(4);
+}
+
+type ResultProcessorNext =
+    unsafe extern "C" fn(*mut Header, res: *mut MaybeUninit<SearchResult>) -> RPStatus;
+type ResultProcessorFree = unsafe extern "C" fn(*mut Header);
+
+/// Properties of Result Processors that are accessed by C code through FFI. This type is named `Header` because it
+/// must be the first member of the [`ResultProcessor`] struct in order to guarantee that we can cast a `*mut ResultProcessor<P>`
+/// to a `*mut Header` (which is what the C code expects to receive).
+///
+/// # Pinning Safety Contract
+///
+/// Result processors intrusively linked, where one result processor has the pointer to the
+/// previous (forming an intrusively singly-linked list). This places a big safety invariant on all code
+/// that touches this data structure: Result processors must *never* be moved while part of a QueryIterator
+/// chain. Accidentally moving a processor will remove in a broken link and undefined behaviour.
+///
+/// This invariant is implicit in the C code where its uncommon to move the heap allocated objects (which would
+/// mean memcopying them around). Rust move semantics make it very easy to break this invariant however, which is why
+/// all APIs will only ever return either immutable or pinned mutable references to the data structure.
+///
+/// This crate has to ensure that these pinning invariants are upheld internally, but when sending pointers through the FFI boundary
+/// this is impossible to guarantee. Thankfully, as mentioned above its quite rare to move out of pointers, so its reasonably safe.
+#[repr(C)]
+#[derive(Debug)]
+pub struct Header {
+    /// Reference to the parent structure
+    pub(super) parent: *mut c_void, // QueryIterator*
+
+    /// Previous result processor in the chain
+    pub(super) upstream: *mut Self,
+
+    /// Type of result processor
+    ty: ResultProcessorType,
+
+    /// time measurements
+    giltime: timespec,
+
+    /// "VTable" function. Pulls SearchResults out of this result processor.
+    pub(super) next: Option<ResultProcessorNext>,
+    /// "VTable" function. Destroys and deallocates the result processor. This is only ever called by C code.
+    free: Option<ResultProcessorFree>,
+
+    // ResultProcessors *must* be !Unpin to ensure they are not moved about breaking the linked list.
+    _unpin: PhantomPinned,
+}
+
+pin_project! {
+    #[derive(Debug)]
+    #[repr(C)]
+    pub struct ResultProcessor<P> {
```

Comment:
Style comment/question: can we avoid using the same name for both the trait and the struct? 🤔

Reply:
I agree it's unfortunate, do you have a better name idea? `FFIResultProcessorWrapper`?

Reply:
`ResultProcessorEnvelope`? I don't know 🤷🏻

Reply:
Chose `ResultProcessorWrapper` should be alright